### PR TITLE
feature: backreference subjects - breaking change

### PR DESCRIPTION
**Pull request by colek42**
The PR adds support for traversing the attestation graph by specifying "backref" subjects that link steps of the CI process.

### CODE REVIEW

**Comment by colek42** [comment]
@flickerfly this is the last MR we need for out end to end gitlab demo https://gitlab.com/testifysec/demos/witness-demo/-/pipelines/488481541

**Comment by colek42** [comment]
@mikhailswift I wonder if it makes sense to define backref subjects and the max search depth in the policy?

**Comment by flickerfly** [comment]
A suggestion, I get why witness is ideally in the environment that each task runs it, but that comes with some heavy lifting. Could an example of a less strong, but easier starting point pipeline be made available that does some of the same work at the end of the pipeline run by picking up artifacts that were generated in previous tasks? This way we could just add a task with an image and get some of the benefit without re-tooling the entire pipeline to include witness, then maybe ease it into a better position. (I'm thinking strangler pattern.)

**Comment by colek42** [comment]
@flickerfly we are working on tooling to help automate instrumentation into GitLab.  

Adding witness to just the build and packaging (sometimes this is a single step) steps will get you to SLSA level 3.  Happy to discuss specific use cases.

**Review by mikhailswift — APPROVED** [review]

**Comment by mikhailswift** [comment]
conflicts on this branch

File: cmd/witness/cmd/verify.go [hunk 1]
```diff
@@ -16,16 +16,18 @@ package cmd
...
+	"crypto/sha256"
 	"encoding/json"
 	"fmt"
 	"io"
 	"os"
 
 	"github.com/spf13/cobra"
 	"github.com/testifysec/witness/cmd/witness/options"
-	"github.com/testifysec/witness/pkg"
+	witness "github.com/testifysec/witness/pkg"
```
Comment (colek42) [inline cmd/witness/cmd/verify.go:27]: go fmt does this

**Review by mikhailswift — APPROVED** [review]

**Pull request merged**